### ISSUE TITLE
Loss.backward() changes target values

### ISSUE THREAD

**Issue by wboerdijk** [pytorch/pytorch#39942]
I am training a custom segmentation network with standard `CrossEntropyLoss()` and three target classes. With a batch size of 1, after 36 passes the backward pass seems to change the values of the target tensor. 
To validate this I print the unique values of the tensor (with `torch.unique`) before calling `loss.backward()`. Results are **always the same**, and **always after 36 passes**:

```
# before backward / directly after loss calculation: tensor([0, 1, 2])
# after backward: tensor([-5340268357666996224, -5340268354615407047, 0, 1, 2, 3051589177])
```
The target doesn't `require_grad`. Strangely, this happens regardless of 
- learning rate
- optimizer
- shuffling the data or not (doesn't matter which ground truth I take, I always get the same result at 36 passes)
- network weight initialization / pretrained state dict

I tried running the code on another PC with the same result. Running with `CUDA_LAUNCH_BLOCKING=1` doesn't provide any other output, as the script fails later during a calculation of intersection over union (Otherwise I wouldn't have noticed this anyways).

Unfortunately I don't have a small working example to reproduce this behavior, since this is part of a larger project. I'd hope that someone could point me in a direction where I should look to further investigate this, and/or an answer to why this might happen at all, and/or other stuff I should have checked. I have been using this project for a while now (including data loading, network, loss calculation etc.) and everything has been working fine.

Other information:
- PyTorch version: 1.5.0
- CUDA: 10.2
- cudatoolkit: 10.2.89

If any other information is neccessary I'll provide it asap!

**Comment by agolynski** [comment]
loss.backward() should not change data in the tensors themselves, e.g. see https://discuss.pytorch.org/t/what-does-the-backward-function-do/9944 even if target has requires_grad.

why running 36 passes at once, what if you run just one pass at a time and track when data of your underlying tensor changes?

does this behavior change if run on CPU?

**Comment by wboerdijk** [comment]
Thanks for your quick reply @agolynski !

> why running 36 passes at once, what if you run just one pass at a time and track when data of your underlying tensor changes?

That is actually what I am doing, sorry for the bad explanation. I do one forward pass, calculate loss, and backpropagate. After doing this 36 times the target tensor gets changed suddenly - irrespective of the current data I am using etc.

> does this behavior change if run on CPU?

I couldn't test this as I am using a correlation layer from [Nvidia's flownet](https://github.com/NVIDIA/flownet2-pytorch). This only supports cuda by default, I haven't implemented a cpu version.

**Comment by agolynski** [comment]
just to make sure: this target tensor is only used in computations and you never update it in the learning loop? 

judging by error messages, seems like you have some sort of memory corruption. to make sure, could you change the size of the tensor without affecting functionality of the program (maybe add a bogus dimension which you'd ignore if possible)?

**Comment by wboerdijk** [comment]
I am actually storing the tensors in a trainer class to have them available for later, and that might have caused this issue. After forwarding I save the tensor in a storage like 
```
self._storage = data_sample
```
to have them accessible for some other methods (metrics, writers, ...). Now, after reading your answer I thought about initializing the storage before each forward like 
```
self._storage = None

...

# load sample, forward, store
```
and now it runs perfectly fine. I am a bit confused as of why this is (I don't have any try...excepts that would have left targets from the previous iteration there if the forwarding would have failed (which it also doesn't, I checked that), and there is no error pointing at an empty storage now), but your thoughts on the memory corruption seem valid to me. Maybe there were still some references to the previous tensor?

Anyways, thanks a bunch for your quick and helpful replies @agolynski , I highly appreciate it! For me this issue is solved, I am closing it!
